### PR TITLE
Introduce fgDebugCheckTypes

### PR DESCRIPTION
**Pull request by EgorBo**
- Validate types used in IR nodes.
- Remove `varTypeIsSmallInt` because it's the same as `varTypeIsSmall`.
- Remove `varTypeGCtype` because it's the same as `varTypeGC`
- Fix small-typed nodes produced by optOptimizeBools (e.g. `ubyte GT_OR(op1, op2) ...`)
- Fix small-typed nodes used by GT_HWINTRINSIC, e.g. all comparisons and ToScalar

### CODE REVIEW

**Comment by EgorBo** [comment]
the latest diffs are better https://dev.azure.com/dnceng-public/public/_build/results?buildId=467438&view=ms.vss-build-web.run-extensions-tab but there are still regressions because of GT_HWINTRINSIC (mainly around op_Equality)

**Comment by EgorBo** [comment]
@jakobbotsch @dotnet/jit-contrib PTAL

[Small diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=479415&view=ms.vss-build-web.run-extensions-tab) are caused by the `GT_HWINTRINSIC` changes (there were qute a few trees with small types as IR types)

Size regressions on arm64 seem to be perf improvements - optOptimizeBool kicks in where previously it didn't, e.g.:
image
(less branches, but larger size), in IR it's:

Main:

```
***** BB03
N006 (  9, 10) [000032] -----------                         *  JTRUE     void   $VN.Void
N005 (  7,  8) [000031] J------N---                         \--*  NE        int    $182
N003 (  5,  5) [000207] -----------                            +--*  HWINTRINSIC ubyte  int op_Equality $142
N001 (  1,  2) [000206] -----------                            |  +--*  CNS_VEC   simd16<0x00000000, 0x00000000, 0x00000000, 0x00000000> $102
N002 (  3,  2) [000284] -----------                            |  \--*  CNS_VEC   simd16<0x00000003, 0x00000003, 0x00000003, 0x00000003> $100
N004 (  1,  2) [000030] -----------                            \--*  CNS_INT   int    0 $40


N010 ( 17, 16) [000025] -----------                         *  JTRUE     void   $VN.Void
N009 ( 15, 14) [000024] J------N---                         \--*  NE        int   
N007 ( 13, 11) [000295] -----------                            +--*  OR        int   
N003 (  7,  5) [000177] -----------                            |  +--*  HWINTRINSIC int    int op_Equality $142
N001 (  3,  2) [000288] -----------                            |  |  +--*  CNS_VEC   simd16<0x00000005, 0x00000005, 0x00000005, 0x00000005> $101
N002 (  3,  2) [000287] -----------                            |  |  \--*  CNS_VEC   simd16<0x00000003, 0x00000003, 0x00000003, 0x00000003> $100
N006 (  5,  5) [000209] -----------                            |  \--*  HWINTRINSIC int    int op_Equality $144
```

PR:
```
***** BB02
N006 ( 11, 10) [000025] -----------                         *  JTRUE     void   $VN.Void
N005 (  9,  8) [000024] J------N---                         \--*  NE        int    $181
N003 (  7,  5) [000176] -----------                            +--*  HWINTRINSIC ubyte  int op_Equality $141
N001 (  3,  2) [000283] -----------                            |  +--*  CNS_VEC   simd16<0x00000005, 0x00000005, 0x00000005, 0x00000005> $101
N002 (  3,  2) [000282] -----------                            |  \--*  CNS_VEC   simd16<0x00000003, 0x00000003, 0x00000003, 0x00000003> $100
N004 (  1,  2) [000023] -----------                            \--*  CNS_INT   int    0 $40
```

File: src/coreclr/jit/assertionprop.cpp [hunk 1]
```diff
@@ -232,25 +232,75 @@ bool IntegralRange::Contains(int64_t value) const
...
 #else
 #error Unsupported platform
 #endif
-                    // TODO-Casts: specify more precise ranges once "IntegralRange" supports them.
-                    return {SymbolicIntegerValue::Zero, SymbolicIntegerValue::ByteMax};
                 default:
+                    if (varTypeIsSmall(node->AsHWIntrinsic()->GetSimdBaseType()))
+                    {
+                        return ForType(node->AsHWIntrinsic()->GetSimdBaseType());
+                    }
```
[the remark this answers is not in the document]
Comment (EgorBo) [reply src/coreclr/jit/assertionprop.cpp:303]: I can probably just list them directly indeed

File: src/coreclr/jit/fgdiagnostic.cpp [hunk 2]
```diff
@@ -3239,6 +3239,56 @@ void Compiler::fgDebugCheckBBlist(bool checkBBNum /* = false */, bool checkBBRef
...
 
+//------------------------------------------------------------------------
+// fgDebugCheckTypes: Validate node types used in the given tree
+//
+// Arguments:
+//    tree - the tree to (recursively) check types for
+//
+void Compiler::fgDebugCheckTypes(GenTree* tree)
+{
+    fgWalkTreePost(&tree, [](GenTree** pNode, fgWalkData* data) -> fgWalkResult {
```
[the remark this answers is not in the document]
Comment (EgorBo) [reply src/coreclr/jit/fgdiagnostic.cpp:3250]: sure thing

**Review by jakobbotsch — APPROVED** [review]

**Comment by BruceForstall** [comment]
Thanks for doing this! I'm looking forward to all the follow-up work on the TODOs :-)

**Pull request merged**